src/java_bytecode/java_object_factory.cpp

```diff
@@ -498,6 +509,79 @@ class recursion_set_entryt
   }
 };
 
+/// Generic-recursion-map entries owner class. If a generic-recursion-map
+/// entries are added in a particular scope, ensures that they is erased on
+/// leaving that scope.
+class generic_recursion_map_keyst
+{
+  /// Recursion map to modify
+  std::unordered_map<irep_idt, std::stack<reference_typet>, irep_id_hash>
+    &generic_recursion_map;
+  /// Keys of the entries to pop on destruction
+  std::vector<irep_idt> erase_keys;
+
+public:
+  /// Initialize a generic-recursion-map entry owner operating on a given map.
+  /// Initially it does not own any map entry.
+  /// \param _generic_recursion_map: map to operate on.
+  explicit generic_recursion_map_keyst(
+    std::unordered_map<irep_idt, std::stack<reference_typet>, irep_id_hash>
+      &_generic_recursion_map)
+    : generic_recursion_map(_generic_recursion_map)
+  {
+  }
+
+  /// Removes the top of the stack for each key in erase_keys from the
+  /// controlled map.
+  ~generic_recursion_map_keyst()
+  {
+    for(const auto key : erase_keys)
+    {
+      PRECONDITION(generic_recursion_map.count(key));
+      (*generic_recursion_map.find(key)).second.pop();
+      if((*generic_recursion_map.find(key)).second.empty())
+      {
+        generic_recursion_map.erase(key);
+      }
+    }
+  }
+
+  generic_recursion_map_keyst(const generic_recursion_map_keyst &) = delete;
+  generic_recursion_map_keyst &
+  operator=(const generic_recursion_map_keyst &) = delete;
+
+  /// Add pairs to the controlled map. Own the keys and pop from their stack
+  /// on destruction; otherwise do nothing.
+  /// \param parameters generic parameters that are the keys of the pairs to add
+  /// \param types a type to add for each parameter
+  const void insert_pairs(
+    const std::vector<java_generic_parametert> &parameters,
+    const std::vector<reference_typet> &types)
+  {
+    INVARIANT(erase_keys.empty(), "insert_pairs should only be called once");
+    PRECONDITION((parameters.size() == types.size()) && parameters.size() > 0);
+    for(size_t i = 0; i < parameters.size(); i++)
```

Comment:
why not ranged for over `parameters` ?

Reply:
I need the index of the current parameter to retrieve the type at the same index from `types`.